presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ColumnRangesSystemTable.java

```diff
@@ -0,0 +1,188 @@
+        implements SystemTable
+{
+    private static final String MIN_COLUMN_SUFFIX = "_min";
+    private static final String MAX_COLUMN_SUFFIX = "_max";
+    private static final String COLUMN_RANGES_TABLE_SUFFIX = "$column_ranges";
+
+    private final IDBI dbi;
+    private final RaptorTableHandle sourceTable;
+    private final List<TableColumn> indexedRaptorColumns;
+    private final ConnectorTableMetadata tableMetadata;
+
+    public ColumnRangesSystemTable(RaptorTableHandle sourceTable, IDBI dbi)
+    {
+        this.sourceTable = requireNonNull(sourceTable, "sourceTable is null");
+        this.dbi = requireNonNull(dbi, "dbi is null");
+
+        this.indexedRaptorColumns = dbi.onDemand(MetadataDao.class).listTableColumns(sourceTable.getTableId()).stream()
+                .filter(column -> isIndexedType(column.getDataType()))
+                .collect(toImmutableList());
+        List<ColumnMetadata> systemTableColumns = this.indexedRaptorColumns.stream()
+                .flatMap(column -> Stream.of(
+                        new ColumnMetadata(column.getColumnName() + MIN_COLUMN_SUFFIX, column.getDataType(), null, false),
+                        new ColumnMetadata(column.getColumnName() + MAX_COLUMN_SUFFIX, column.getDataType(), null, false)))
+                .collect(toImmutableList());
+        SchemaTableName tableName = new SchemaTableName(sourceTable.getSchemaName(), sourceTable.getTableName() + COLUMN_RANGES_TABLE_SUFFIX);
+        this.tableMetadata = new ConnectorTableMetadata(tableName, systemTableColumns);
+    }
+
+    public static Optional<SchemaTableName> tryGetSourceTable(SchemaTableName tableName)
+    {
+        if (isColumnRangesTable(tableName)) {
+            return Optional.of(new SchemaTableName(
+                tableName.getSchemaName(),
+                    tableName.getTableName().substring(0, tableName.getTableName().length() - COLUMN_RANGES_TABLE_SUFFIX.length())));
+        }
+        return Optional.empty();
+    }
+
+    private static boolean isColumnRangesTable(SchemaTableName tableName)
+    {
+        return tableName.getTableName().endsWith(COLUMN_RANGES_TABLE_SUFFIX) && tableName.getTableName().length() > COLUMN_RANGES_TABLE_SUFFIX.length();
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return tableMetadata;
+    }
+
+    @Override
+    public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        String metadataSqlQuery = getColumnRangesMetadataSqlQuery(sourceTable, indexedRaptorColumns);
+        List<Type> columnTypes = tableMetadata.getColumns().stream()
+                .map(ColumnMetadata::getType)
+                .collect(toImmutableList());
+
+        ResultSetValues resultSetValues = new ResultSetValues(columnTypes);
```

Comment:
This class is a bit weird as it's designed to be used directly for implementing `RecordCursor`. Does using it here buy us anything? I think it'd be simpler to just handle `ResultSet` directly, especially since we already need the Java type checks below.

Reply:
It might be easier to change this to implement `pageSource` rather than `cursor`, then we can just build a single row page directly by looping over the fields.

Reply:
Agreed. It doesn't make sense to use this class here since we only have a single row. Not sure why I used it before. Will also change to the `pageSouce` implementation.